File: scripts/Matlab/plot_csv.m
function plot_csv(varargin)
% PLOT_CSV  Grafica uno o varios .csv generados por bag2csv.py (o por
%   bag2csv_crazyflie.py): una figura por fichero, una curva por columna
%   numérica frente a Timestamp (convertido a segundos relativos al
%   inicio del registro).
%
%   Complementa a Crazyflie_Graphs.m: ese script espera variables ya
%   cargadas en el workspace (references, dron_state, ...) generadas por
%   el pipeline Crazyflie_Launch.m + bag2txt_sim.sh + Txt2Mat.m. Este
%   lee directamente los .csv de bag2csv.py, sin ese pipeline intermedio.
%
%   Uso:
%       plot_csv('fichero1.csv')
%       plot_csv('fichero1.csv', 'fichero2.csv')

    if nargin < 1
        error('plot_csv: hay que pasar al menos un fichero .csv');
    end

    for k = 1:nargin
        file = varargin{k};
        data = readtable(file);

        if height(data) == 0
            warning('%s: vacío, se omite', file);
            continue
        end

        t0 = data.Timestamp(1);
        t = (data.Timestamp - t0) / 1e9; % ns -> s

        value_vars = data.Properties.VariableNames(2:end);
        numeric_vars = value_vars(varfun(@isnumeric, data(:, value_vars), ...
            'OutputFormat', 'uniform'));

        if isempty(numeric_vars)
            warning('%s: no tiene columnas numéricas que graficar', file);
            continue
        end

        [~, name, ext] = fileparts(file);
        figure('Name', [name ext]);
        hold on
        for v = 1:numel(numeric_vars)
            plot(t, data.(numeric_vars{v}));
        end
        hold off
        xlabel('Tiempo [s]')
        title([name ext], 'Interpreter', 'none')
        legend(numeric_vars, 'Interpreter', 'none', 'Location', 'best')
        grid minor
    end
end

File: scripts/Python/plot_csv.py
# -*- coding: utf-8 -*-
# Grafica directamente con matplotlib uno o varios .csv generados por
# bag2csv.py (o por bag2csv_crazyflie.py): una figura por fichero, una
# curva por columna numérica frente a Timestamp. Pensado para una
# inspección rápida sin tener que abrir Matlab.
#
# Uso: python3 plot_csv.py fichero1.csv [fichero2.csv ...]

import csv
import sys

import matplotlib.pyplot as plt


def plot_file(path):
    with open(path, newline='') as f:
        reader = csv.reader(f)
        header = next(reader)
        rows = list(reader)

    if not rows:
        print('%s: vacío, se omite' % path)
        return

    t0 = float(rows[0][0])
    timestamps = [(float(r[0]) - t0) / 1e9 for r in rows]  # ns -> s, relativo al inicio

    numeric_cols = []
    for col in range(1, len(header)):
        try:
            float(rows[0][col])
            numeric_cols.append(col)
        except ValueError:
            pass

    if not numeric_cols:
        print('%s: no tiene columnas numéricas que graficar (¿es un topic de texto?)' % path)
        return

    plt.figure(path)
    for col in numeric_cols:
        values = [float(r[col]) for r in rows]
        plt.plot(timestamps, values, label=header[col])
    plt.xlabel('Tiempo [s]')
    plt.title(path)
    plt.legend(loc='best')
    plt.grid(True, which='both', linestyle=':')


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print('Uso: python3 plot_csv.py fichero1.csv [fichero2.csv ...]')
        sys.exit(1)

    for csv_path in sys.argv[1:]:
        plot_file(csv_path)

    plt.show()
